src/js/libcs/OpSound.js

```diff
@@ -1,48 +1,403 @@
-
             erg = evaluate(modifs.line);
             if (erg.ctype === 'number') {
-                linenumber = Math.floor(erg.value.real);
-                if (linenumber < 0) {
-                    linenumber = 0;
-                }
-                if (linenumber > 10) {
-                    linenumber = 10;
-                }
+                var val = erg.value.real;
+            }
+            else if (erg.ctype === 'string'){
+              var val = erg.value;
+            }
+            else {
+                var val = 0;
+            }
+            if (val === 0) {
+              line = 0;
+            }
+            else if (!linesDict[val]){
+              line = linesCounter;
+              linesDict[val] = line;
+              linesCounter++;
+            }
+            else {
+              line = linesDict[val];
+            }
+        }
+        if (modifs.partials !== undefined) {
+          erg = evaluate(modifs.partials);
+          if (erg.ctype === 'list') {
+              partials = erg.value;
+          }
+        }
+    }
+    //should i read lists as cslist or jslist?
+    var freq = evaluate(args[0]).value.real;
+    var amp = 0.5;
+    var damp = 0;
+    var duration = 1;
+    var harmonics = List.asList(CSNumber.real(1)).value;
+    var line = 0;
+    var partials = Array(harmonics.length).fill(CSNumber.real(1));
+    var attack = 0.01;
+    var release = 0.01;
+
+    handleModifs();
+    if (harmonics.length>partials.length){
+      partials = Array(harmonics.length).fill(CSNumber.real(1))
+    }
+
+
+    function playOscillator(line, freq, gain) {
+      var oscNode = audioCtx.createOscillator();
+      oscNode.type = 'sine';
+      oscNode.frequency.value = freq;
+      var gainNode = audioCtx.createGain();
+      gainNode.connect(lines[line].masterGain);
+      gainNode.gain.value = 0;
+      oscNode.connect(gainNode);
+      //gainNode.gain.value = gain;
+      gainNode.gain.linearRampToValueAtTime(gain, audioCtx.currentTime+attack);
+      oscNode.start(0);
+      oscNode.onended = function () {//maybe delete
+        //gainNode.disconnect();
+        //gainNode = null;
+      }
+      if(duration >= 0){
+        gainNode.gain.setValueAtTime(0,audioCtx.currentTime + duration+attack);
+
+      }
+      return {oscNode: oscNode, gainNode: gainNode};
+    }
+
+    if(lines[line] === 0){ //initialize
+      var masterGain = audioCtx.createGain();
+      masterGain.gain.value = 0;
+      lines[line] = {oscNodes: [], masterGain: masterGain};
+      for(var i=0; i < harmonics.length; i++){
+        lines[line].oscNodes.push(playOscillator(line, partials[i].value.real*(i+1)*freq, harmonics[i].value.real));
+      }
+      masterGain.gain.linearRampToValueAtTime(amp, audioCtx.currentTime+attack);
+      masterGain.connect(audioCtx.destination);
+      if(damp > 0){ //not properly mathematically but does the job
+        masterGain.gain.setTargetAtTime(0.0, audioCtx.currentTime+attack,(1/damp));
```

Comment:
maybe it is worth to try to stop here all the oscillating nodes after 6*(1/damp) seconds, because then the sound will be unhearable anyway (e^-6<1/256).

Reply:
tried this, but throws random errors, probably because, if it is stopped elsewhere, it can't be stopped again. I dont think this makes a big difference anyways.